module/zfs/zpl_xattr.c

```diff
@@ -1408,6 +1397,677 @@ xattr_handler_t zpl_xattr_acl_default_handler =
 
 #endif /* CONFIG_FS_POSIX_ACL */
 
+#if (defined(CONFIG_FS_POSIX_ACL) && !defined(HAVE_GET_ACL) && \
+	!defined(HAVE_CHECK_ACL)) || defined(ZFS_NFS4_ACL)
+int
+_zpl_permission(struct inode *ip, int mask)
+{
+	if (ITOZSB(ip)->z_acl_type == ZFS_ACLTYPE_NFS4ACL) {
+		/*
+		 * XXX - mask could also include
+		 * MAY_APPEND|MAY_ACCESS|MAY_OPEN|MAY_CHDIR, do we care?
+		 * What about V_APPEND zfs_access flag?
+		 */
+		return (-zfs_access(ip, (mask &
+		    (MAY_READ|MAY_WRITE|MAY_EXEC)) << 6, 0, CRED()));
+	} else {
+#if defined(HAVE_CHECK_ACL) || defined(HAVE_GET_ACL)
+		return (generic_permission(ip, mask));
+#else
+		return (generic_permission(ip, mask, __zpl_check_acl));
+#endif
+	}
+}
+
+#if defined(HAVE_PERMISSION_WITH_NAMEIDATA)
+int
+zpl_permission(struct inode *ip, int mask, struct nameidata *nd)
+{
+	return (_zpl_permission(ip, mask));
+}
+#elif defined(HAVE_PERMISSION)
+int
+zpl_permission(struct inode *ip, int mask)
+{
+	return (_zpl_permission(ip, mask));
+}
+#endif /* HAVE_PERMISSION */
+#endif
+/* (CONFIG_FS_POSIX_ACL & !HAVE_GET_ACL & !HAVE_CHECK_ACL) | ZFS_NFS4_ACL */
+
+#ifdef ZFS_NFS4_ACL
+
+static const struct cred *nfs4acl_resolver_cred;
+
+#define	NFS4ACL_XATTR	"system.nfs4_acl"
+#define	NFS4ACL_NAMESZ	128
+#define	NFS4ACL_UINT_MAXLEN	11
+#define	NFS4ACL_XDR_MOD	4
+
+#if defined(HAVE_GENERIC_KEY_INSTANTIATE)
+static int
+nfs4acl_key_preparse(struct key_preparsed_payload *kpp)
+{
+	struct user_key_payload *ukp;
+	size_t dl = kpp->datalen;
+
+	if (kpp->data == NULL || dl < 1 || dl > 32767)
+		return (-EINVAL);
+
+	ukp = kmem_alloc(dl + sizeof (*ukp), KM_SLEEP);
+	ukp->datalen = dl;
+	kpp->quotalen = dl;
+	kpp->payload.data[0] = ukp;
+	memcpy(ukp->data, kpp->data, dl);
+
+	return (0);
+}
+
+static void
+nfs4acl_key_free_preparse(struct key_preparsed_payload *kpp)
+{
+	kfree(kpp->payload.data[0]);
+}
+#else /* HAVE_GENERIC_KEY_INSTANTIATE */
+static int
+nfs4acl_key_instantiate(struct key *k, struct key_preparsed_payload *kpp)
+{
+	struct user_key_payload *ukp;
+	size_t datalen = kpp->datalen;
+	int ret;
+
+	if (!kpp->data || datalen < 1 || datalen > 32767)
+		return (-EINVAL);
+
+	ret = key_payload_reserve(k, datalen);
+	if (ret < 0)
+		return (ret);
+
+	ukp = kmalloc(sizeof (*ukp) + datalen, GFP_KERNEL);
```

Comment:
nit: For consistency with the rest of the ZFS code please use `kmem_alloc()` and `kmem_free()`.  When called with `KM_SLEEP` the allocation cannot fail so you should remove the error handling too.  

```c
        ukp = kmem_alloc(sizeof (*ukp) + datalen, KM_SLEEP);
```

Reply:
D'oh, sorry, you mentioned this before but I forgot when I went back to make it work with the Centos 7 kernel.